File: 08-trees/02_invert_tree.c
#include "../test_runner.h"

typedef struct TreeNode {
    int val;
    struct TreeNode *left, *right;
} TreeNode;

TreeNode* make_tnode(int val) {
    TreeNode* n = malloc(sizeof(TreeNode));
    n->val = val; n->left = NULL; n->right = NULL;
    return n;
}

TreeNode* invert_tree(TreeNode* root) {
    return NULL;
}

int main() {
    TreeNode* r = make_tnode(4);
    r->left = make_tnode(2); r->right = make_tnode(7);
    r->left->left = make_tnode(1); r->left->right = make_tnode(3);
    TreeNode* inv = invert_tree(r);
    TEST_INT("root", inv->val, 4);
    TEST_INT("swapped left", inv->left->val, 7);
    TEST_INT("swapped right", inv->right->val, 2);
    TEST_SUMMARY();
}

File: core/08-trees/02_invert_tree.py
import os, sys
_d = os.path.dirname(os.path.abspath(__file__))
while _d and not os.path.exists(os.path.join(_d, 'test_runner.py')):
    _d = os.path.dirname(_d)
sys.path.insert(0, _d)
from test_runner import run_tests
from collections import deque


# Invert Binary Tree (LC #226) -- Easy
# Input:  root: TreeNode (or None)
# Output: TreeNode (root of mirrored tree)
# Mirror a binary tree.
#
#   Key: Swap left and right children recursively

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def make_tree(arr):
    """Build tree from level-order array. None = no node."""
    if not arr or arr[0] is None:
        return None
    root = TreeNode(arr[0])
    queue = deque([root])
    i = 1
    while queue and i < len(arr):
        node = queue.popleft()
        if i < len(arr) and arr[i] is not None:
            node.left = TreeNode(arr[i])
            queue.append(node.left)
        i += 1
        if i < len(arr) and arr[i] is not None:
            node.right = TreeNode(arr[i])
            queue.append(node.right)
        i += 1
    return root


def invert_tree(root):
    pass


def tree_to_list(root):
    if not root:
        return []
    result = []
    queue = deque([root])
    while queue:
        node = queue.popleft()
        if node:
            result.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append(None)
    while result and result[-1] is None:
        result.pop()
    return result


def solve(arr):
    return tree_to_list(invert_tree(make_tree(arr)))


run_tests(solve, [
    (([4,2,7,1,3,6,9],), [4,7,2,9,6,3,1]),
    (([2,1,3],),          [2,3,1]),
    (([],),               []),
])
